### ISSUE TITLE
Training issues

### ISSUE THREAD

**Issue by mhenrichsen** [neuphonic/neutts-air#98]
Hi

I've compiled a dataset of 3 million danish audio files and have tried to train your Nano model on it. 

The results were unforunately not very good as the model didn't manage to speak a single word correctly.

Here is the training run:
https://wandb.ai/mhenrichsen/tts/runs/9tf9s60a?nw=nwusermhenrichsen

As you see the loss plateus around 7.2.

Here is the dataset:
https://huggingface.co/datasets/syvai/neutts-danish-pretokenized

Which is tokenized from:
https://huggingface.co/datasets/syvai/nota-neucodec and a few other datasets.

I've tried using both your training script and using Axolotl, but results remain the same.

**Comment by harryjulian** [comment]
Hey @mhenrichsen! Just having a quick look, the loss seems quite high, but looking at the dataset it seems that's because the code sequences are quite long. 

The base model was trained on utterances up to 30 seconds (1500 codes), so this could be part of the problem. On this, I'm not sure how such a small model would deal with longer context lengths. 

EDIT: it could be worth double-checking your tokenization/inference scripts are using the correct phonemizer -- I've tripped myself up numerous times here.

**Comment by mhenrichsen** [comment]
Hi. Thanks for taking a look. The dataset is filtered at training time to cap out at 2048 tokens, so that should not be an issue.

**Comment by mhenrichsen** [comment]
Would you mind explicitly telling which pheonomizer is used, so I can check against my data?

**Comment by harryjulian** [comment]
You can see it [here](https://github.com/neuphonic/neutts/blob/d62412dd20e17a204f17ab8d779569662dde0c76/neutts/neutts.py#L102), you'll need to alter it to use the `da` [voice from eSpeak](https://espeak.sourceforge.net/languages.html) if you haven't already.

**Comment by mhenrichsen** [comment]
I’m 99% certain that I’ve used the da voice, but will recheck. If I remember correctly though it did result in some phonemes being “?” 

I’m also not sure if all of the da phonemes are represented in the vocab of the model.

**Comment by mhenrichsen** [comment]
I just compared my tokenization script to yours and they seem similar, except for some padding (which axolotl adds) and some filtering.

https://gist.github.com/mhenrichsen/11a44bea9e7cbbb175c82d3998c658cb

```
 tokenize_for_axolotl.py: All 1s (no padding)                                                 
  finetune.py: (input_ids != pad_token_id)                                                     
  Data Filtering:                                                                              
  - tokenize_for_axolotl.py: Only filters by token count (min_tokens/max_tokens) and empty text
  - finetune.py: More aggressive filtering that rejects:                                       
    - Text containing digits                                                                   
    - Acronyms (e.g., "U.S.A." or "NASA")                                                      
    - Text not ending in .,?!                                                                  
    - Text containing £ or $                                                                   
                                                                                               
  Phoneme processing is identical in both: " ".join(phones[0].split())                         
                                                                                               
  Label masking logic is the same: labels set to -100 before <|SPEECH_GENERATION_START|>.      
                                                                                               
  The most significant functional difference is that finetune.py uses fixed-length padded      
  sequences suitable for batched training with HuggingFace Trainer, while                      
  tokenize_for_axolotl.py outputs variable-length sequences for Axolotl (which handles         
  padding/batching itself).
```
Edit: can you share what loss your would expect to see?
